File: src/generation/RandomInputGenerator.cpp
#include "generation/RandomInputGenerator.hpp"
#include <algorithm>
#include <chrono>
#include <cmath>
#include <stdexcept>

// default constructor:
RandomInputGenerator::RandomInputGenerator()
    :generator(
        static_cast<unsigned int> (
            std::chrono::system_clock::now()
                .time_since_epoch()
                .count()
        )
    ){
}

// parameterized constructor:
RandomInputGenerator::RandomInputGenerator(
    unsigned int seed
)
    :generator(seed){
}

// Random input generation function:
std::vector<int> RandomInputGenerator::generateRandom(
            std:: size_t size,
            int min,
            int max
)
{
    if(min>max){
        throw std::invalid_argument("Minimum value cannot be greater than maximum value.");
    }
    std::vector<int> data(size);
    std::uniform_int_distribution<int> distribution(min, max);
    for(std::size_t i = 0; i < size; ++i){
        data[i] = distribution(generator);
    }
    return data;
}

// Sort data according to requested order:

void RandomInputGenerator::sortData(
    std::vector<int>& data,
    SortOrder order
){
    if(order == SortOrder::ASCENDING){
        std::sort(data.begin(), data.end());
    } else {
        std::sort(data.begin(), data.end(), std::greater<int>());
    }
}

void RandomInputGenerator::changeOrdering(
        std::vector<int>& data,
        double percentage,
        SortOrder order
)
{
    sortData(data, order);
    int no_of_swaps = static_cast<int>(std::round(data.size() * percentage));
    std::uniform_int_distribution<std::size_t> distribution(0, data.size() - 1);
    for(int i = 0; i < no_of_swaps; ++i){
        std::size_t index1 = distribution(generator);
        std::size_t index2 = distribution(generator);
        std::swap(data[index1], data[index2]);
    }
}

std::vector<int> RandomInputGenerator::generateInput(
        std::size_t size,
        int min,
        int max,
        InputType type,
        SortOrder order,
        double percentage
){
    std::vector<int> data = generateRandom(size, min, max);
    switch(type){
        case InputType::SORTED:
            sortData(data, order);
            break;
        case InputType::NEARLY_SORTED:
            changeOrdering(data, 0.15, order);
            break;
        case InputType::HIGHLY_INVERSIONAL:
            changeOrdering(data, 0.8, order == SortOrder::ASCENDING ? SortOrder::DESCENDING : SortOrder::ASCENDING);
            break;
        default:
            throw std::invalid_argument("Invalid input type.");
    }

    return data;
}

File: java/generation/RandomInputGenerator.java
package generation;
import java.util.Arrays;
import java.util.Random;

public class RandomInputGenerator{
    public enum InputType{
        RANDOM, 
        SORTED,
        REVERSE_SORTED,
        NEARLY_SORTED,
        HIGHLY_INVERSIONAL
    }

    public enum SortOrder{
        ASCENDING,
        DESCENDING
    }

    private Random Generator;

    public RandomInputGenerator(){
        this.Generator = new Random();
    }

    private int[] generateRandomArray(int size, int minValue, int maxValue){
        int[] array = new int[size];
        for(int i = 0; i < size; i++){
            array[i] = Generator.nextInt(minValue, maxValue + 1);
        }
        return array;
    }

    private void sortData(int[] array, SortOrder order){
       Arrays.sort(array);
       if(order == SortOrder.DESCENDING){
        reverseArray(array);
    }
    }

    private void reverseArray(int[] array){
        int start = 0;
        int end = array.length - 1;
        while(start < end){
            int temp = array[start];
            array[start] = array[end];
            array[end] = temp;
            start++;
            end--;
        }
    }

    public int[] generateInput(int size, int minValue, int maxValue, InputType type, SortOrder order, double inversionPercentage ){
        int[] array = generateRandomArray(size, minValue, maxValue);
        int numberOfInversions = 0;
        switch(type){ 
            case RANDOM:
                break;
            case SORTED:
                sortData(array, order);
                break;
            case REVERSE_SORTED:
                sortData(array, order == SortOrder.ASCENDING ? SortOrder.DESCENDING : SortOrder.ASCENDING);
                break;
            case NEARLY_SORTED:
                sortData(array, order);
                // Introduce a small number of inversions
                numberOfInversions = (int)(size * inversionPercentage);
                for(int i = 0; i < numberOfInversions; i++){
                    int index1 = Generator.nextInt(size);
                    int index2 = Generator.nextInt(size);
                    int temp = array[index1];
                    array[index1] = array[index2];
                    array[index2] = temp;
                }
                break;
            case HIGHLY_INVERSIONAL:
                numberOfInversions = (int)(size * inversionPercentage);
                sortData(array, order == SortOrder.ASCENDING ? SortOrder.DESCENDING : SortOrder.ASCENDING);
                // Introduce a large number of inversions
                numberOfInversions = (int)(size * inversionPercentage);
                for(int i = 0; i < numberOfInversions; i++){
                    int index1 = Generator.nextInt(size);
                    int index2 = Generator.nextInt(size);
                    int temp = array[index1];
                    array[index1] = array[index2];
                    array[index2] = temp;
                }
                break;
        }
        return array;
    }


}
